store/tikv/conn_pool.go

```diff
@@ -17,132 +17,104 @@
-	poolFunc := func() (pools.Resource, error) {
-		r, err := f(addr)
-		if err == nil && r == nil {
-			return nil, errors.Errorf("cannot create nil connection")
-		}
-		return r, errors.Trace(err)
-	}
-	p := new(Pool)
-	p.p = pools.NewResourcePool(poolFunc, capability, capability, poolIdleTimeoutSeconds*time.Second)
-	return p
-}
-
-// GetConn takes a connection out of the pool.
-func (p *Pool) GetConn() (*Conn, error) {
-	conn, err := p.p.Get()
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	return conn.(*Conn), nil
-}
-
-// PutConn puts a connection back to the pool.
-func (p *Pool) PutConn(c *Conn) {
-	if c == nil {
-		return
-	} else if c.closed {
-		// if c is closed, we will put nil
-		p.p.Put(nil)
-	} else {
-		p.p.Put(c)
-	}
+// Conn is a simple wrapper of grpc.ClientConn.
+type Conn struct {
+	c *grpc.ClientConn
+	// The reference count of how many TiKVClient uses this grpc.ClientConn at present.
+	// TODO Use refCount to impl background idle Conn cleanup.
+	refCount uint32
 }
 
-// Close closes the pool.
-func (p *Pool) Close() {
-	p.p.Close()
-}
+// createConnFunc is the type of functions that can be used to create a grpc.ClientConn.
+type createConnFunc func(addr string) (*grpc.ClientConn, error)
 
-// Pools maintains connections with multiple addrs.
-type Pools struct {
+// ConnPool is a pool that maintains Conn with a specific addr.
+// TODO Implement background cleanup: add a backgroud goroutine to periodically check
+// whether there is any Conn in pool have no usage (refCount == 0), and then
+// close and remove it from pool.
+type ConnPool struct {
 	m struct {
-		sync.Mutex
-		isClosed   bool
-		capability int
-		mpools     map[string]*Pool
+		sync.RWMutex
+		isClosed bool
+		conns    map[string]*Conn
 	}
 	f createConnFunc
 }
 
-// NewPools creates a Pools. It maintains a Pool for each address, and each Pool
-// has the same capability.
-func NewPools(capability int, f createConnFunc) *Pools {
-	p := new(Pools)
+// NewConnPool creates a ConnPool.
+func NewConnPool(f createConnFunc) *ConnPool {
+	p := new(ConnPool)
 	p.f = f
-	p.m.capability = capability
-	p.m.mpools = make(map[string]*Pool)
+	p.m.conns = make(map[string]*Conn)
 	return p
 }
 
-// GetConn takes a connection out of the pool by addr.
-func (p *Pools) GetConn(addr string) (*Conn, error) {
-	p.m.Lock()
+// Get takes a Conn out of the pool by the specific addr.
+func (p *ConnPool) Get(addr string) (*grpc.ClientConn, error) {
+	p.m.RLock()
 	if p.m.isClosed {
-		p.m.Unlock()
-		return nil, errors.Errorf("pools is closed")
+		p.m.RUnlock()
+		return nil, errors.Errorf("ConnPool is closed")
 	}
-	pool, ok := p.m.mpools[addr]
+	conn, ok := p.m.conns[addr]
+	p.m.RUnlock()
 	if !ok {
-		pool = NewPool(addr, p.m.capability, p.f)
-		p.m.mpools[addr] = pool
+		var err error
+		conn, err = p.tryCreate(addr)
+		if err != nil {
+			return nil, err
+		}
 	}
-	p.m.Unlock()
-
-	return pool.GetConn()
+	// Increment refCount.
+	atomic.AddUint32(&conn.refCount, 1)
```

Comment:
I think you need Mutex to protect _refCount_. Consider later you want to close _Conn_ when _refCount_ is _0_, the _Conn_ may be closed unexpectedly by other thread before increasing its _refCount_.